megamek/src/megamek/client/bot/princess/HeatMap.java

```diff
@@ -0,0 +1,816 @@
+
+            // Interpolate movement from last known position
+            mapAdjustment = getMovementWeightAdjustment(tracked);
+            List<Coords> path = new ArrayList<>();
+            Coords startPosition = lastPositionCache.get(tracked.getId());
+            if (lastPositionCache.containsKey(tracked.getId())) {
+                path = Coords.intervening(startPosition, position);
+                // Remove the starting point, as it was already counted in the last round
+                path.remove(startPosition);
+            } else {
+                path.add(position);
+            }
+
+            updateTeamMovementMap(path, mapAdjustment);
+
+            // If individual entity tracking is enabled, adjust the entity tracker
+            if (trackIndividualUnits) {
+                updateEntityMap(tracked.getId(), path, mapAdjustment);
+            }
+
+            // Stash this position for movement interpolation in the next round
+            updateLastKnown(tracked.getId(), position);
+        }
+    }
+
+    /**
+     * Update the team activity tracker for a given position
+     * @param position    position to update
+     * @param adjustment  modifier to existing value
+     */
+    private void updateTeamActivityMap(Coords position, int adjustment) {
+        int mapValue = teamActivity.getOrDefault(position, 0) + adjustment;
+        teamActivity.put(position, mapValue);
+    }
+
+    /**
+     * Get the weight of the entity for adjusting the team activity tracker. Higher BV units get a
+     * larger weight, while faster units have a reduced weight.
+     * @param tracked  entity to calculate weight for
+     * @return         weight to modify tracker with
+     */
+    private int getTeamWeightAdjustment (Entity tracked) {
+
+        int bvWeight = tracked.getBvCalculator().retrieveBV();
+        if (bvWeight == -1) {
+            bvWeight = tracked.getInitialBV();
+        }
+
+        bvWeight = Math.max(bvWeight - (100 * Math.max(tracked.getJumpMP(), tracked.getWalkMP())), 1);
+        bvWeight = (int) Math.floor(weightScaling * bvWeight);
+
+        return bvWeight;
+    }
+
+    /**
+     * Apply decay rate to team activity map. Positions that have a tracked entity use a slower,
+     * linear decay rate, while those that do not have a tracked entity use a faster exponential
+     * decay rate
+     * @param game   current game
+     */
+    private void ageTeamActivityMap (Game game) {
+
+        // Get positions of tracked entities with known positions
+        List<Coords> activePositions = new ArrayList<>();
+        for (int curId : lastPositionCache.keySet()) {
+            Entity curEntity = game.getEntity(curId);
+            if (curEntity != null &&
+                    !curEntity.isDestroyed() &&
+                    curEntity.isDeployed() &&
+                    !curEntity.isCarcass() &&
+                    (isFriendlyTeam || curEntity.isVisibleToEnemy() || curEntity.isDetectedByEnemy())) {
```

Comment:
This exact block appears several times; it could be replaced with a function, e.g.

    if (okayToTrack(curEntity)) {

with `okayToTrack` containing this long conditional.